gadgets/trace_lsm/program.bpf.c

```diff
@@ -0,0 +1,222 @@
+	F(file_truncate)            \
+	F(task_alloc)               \
+	F(task_free)                \
+	F(cred_alloc_blank)         \
+	F(cred_free)                \
+	F(cred_prepare)             \
+	F(cred_transfer)            \
+	F(cred_getsecid)            \
+	F(kernel_act_as)            \
+	F(kernel_create_files_as)   \
+	F(kernel_module_request)    \
+	F(kernel_load_data)         \
+	F(kernel_post_load_data)    \
+	F(kernel_read_file)         \
+	F(kernel_post_read_file)    \
+	F(task_fix_setuid)          \
+	F(task_fix_setgid)          \
+	F(task_fix_setgroups)       \
+	F(task_setpgid)             \
+	F(task_getpgid)             \
+	F(task_getsid)              \
+	F(current_getsecid_subj)    \
+	F(task_getsecid_obj)        \
+	F(task_setnice)             \
+	F(task_setioprio)           \
+	F(task_getioprio)           \
+	F(task_prlimit)             \
+	F(task_setrlimit)           \
+	F(task_setscheduler)        \
+	F(task_getscheduler)        \
+	F(task_movememory)          \
+	F(task_kill)                \
+	F(task_prctl)               \
+	F(task_to_inode)            \
+	F(userns_create)            \
+	F(ipc_permission)           \
+	F(ipc_getsecid)             \
+	F(msg_msg_alloc_security)   \
+	F(msg_msg_free_security)    \
+	F(msg_queue_alloc_security) \
+	F(msg_queue_free_security)  \
+	F(msg_queue_associate)      \
+	F(msg_queue_msgctl)         \
+	F(msg_queue_msgsnd)         \
+	F(msg_queue_msgrcv)         \
+	F(shm_alloc_security)       \
+	F(shm_free_security)        \
+	F(shm_associate)            \
+	F(shm_shmctl)               \
+	F(shm_shmat)                \
+	F(sem_alloc_security)       \
+	F(sem_free_security)        \
+	F(sem_associate)            \
+	F(sem_semctl)               \
+	F(sem_semop)                \
+	F(netlink_send)             \
+	F(d_instantiate)            \
+	F(getselfattr)              \
+	F(setselfattr)              \
+	F(getprocattr)              \
+	F(setprocattr)              \
+	F(ismaclabel)               \
+	F(secid_to_secctx)          \
+	F(secctx_to_secid)          \
+	F(release_secctx)           \
+	F(inode_invalidate_secctx)  \
+	F(inode_notifysecctx)       \
+	F(inode_setsecctx)          \
+	F(inode_getsecctx)          \
+	F(locked_down)
+
+#define ENUM_ITEM(name) name,
+
+enum lsm_tracepoint { FOR_EACH_LSM_HOOK(ENUM_ITEM) };
+
+struct event {
+	gadget_mntns_id mntns_id;
+	u32 pid;
+	enum lsm_tracepoint tracepoint;
+	char comm[TASK_COMM_LEN];
+};
+
+GADGET_TRACER_MAP(events, 1024 * 256);
+
+GADGET_TRACER(lsm, events, event);
+
+#define DECLARE_LSM_PARAMETER(name)               \
+	const volatile bool trace_##name = false; \
+	GADGET_PARAM(trace_##name);
+
+const volatile bool trace_all = false;
+GADGET_PARAM(trace_all);
+
+FOR_EACH_LSM_HOOK(DECLARE_LSM_PARAMETER)
+
+#define TRACE_LSM(name)                                                \
+	SEC("lsm/" #name)                                              \
+	int trace_lsm_##name()                                         \
```

Comment:
To be really useful, this gadget should access parameters.

And when it is a well-known type such as `struct path *`  or `struct file *`, print the full path using `include/gadget/filesystem.h`.

I wonder if type matches relocations can help (https://github.com/inspektor-gadget/inspektor-gadget/pull/2542).

But that's for another PR.

Reply:
> I wonder if type matches relocations can help (https://github.com/inspektor-gadget/inspektor-gadget/pull/2542).

Do you mean we can use `bpf_core_type_matches` to find `struct path *` types automatically?

Reply:
I was thinking of something like this:

```c
#define TRACE_LSM(name)                                                        \
	union security_list_options___##name##arg1 {                           \
	    int (* #name)(const struct path *);                                \
	};                                                                     \
	union security_list_options___##name##arg2 {                           \
	    int (* #name)(const struct path *, const struct path *);           \
	};                                                                     \
	                                                                       \
	/* ignore attaching failures for compatibility */                      \
	SEC("lsm/" #name)                                                      \
	int trace_lsm_##name(void *arg1, void *arg2, void *arg3)               \
            struct event *event; /*TODO: alloc event in a map */               \
            char *path_str = NULL;                                             \
            struct path *path = NULL;                                          \
            if (bpf_core_type_matches(security_list_options___##name##arg1)) { \
                path = (struct path *) arg1;                                   \
            }                                                                  \
            if (bpf_core_type_matches(security_list_options___##name##arg2)) { \
                path = (struct path *) arg1;                                   \
            }                                                                  \
            if (path) {                                                        \
                path_str = get_path_str(&fs->pwd);                             \
            }                                                                  \
            if (path_str) {                                                    \
                bpf_probe_read_kernel_str(event->path, MAX_STRING_SIZE, path_str); \
            }                                                                  \
            ...
```

But that would work only for hooks with a single `struct path *` argument (for the arg1 test):
* path_truncate
* path_chroot
* inode_getattr
Or with 2 `struct path *` arguments:
* sb_pivotroot
* move_mount

But others would be missed, so I don't know if it is worth it.

In any cases, it is not for this PR.

Reply:
I haven't much understood the code here, are there structs like `security_list_options___*` in kernel?
We can discuss about this tomorrow.

Reply:
Looking into this: https://lwn.net/Articles/967893/

Reply:
I finally gave up after trying for several hours, I couldn't find any CORE API that can be used to match parameter types.